---
title: "R Notebook"
output: html_notebook
---

```{r}
library(readr)
library(dplyr)
library(tidyr)
library(magrittr)
library(ggrepel)
library(ggpubr)
library(reshape2)
library(data.table)
library(tcR)
library(ggplot2)
library(forcats)

```


```{r}
library(ggplot2)
library(stringr)

source("/raid/users/pk/kmer_dig/kmer_full_withV.R")

directory="/raid/users/pk/working_dir/mixcr_results_from_KK_CD8/"
k_mer_files <- dir(directory)
k_mer_files

`%+%` <- function(a, b) paste0(a, b)

name1 <- grep(".+_PB.+k_mers.txt", k_mer_files, value=TRUE)
name1
name2 <- grep(".+_SF.+k_mers.txt", k_mer_files, value=TRUE)
name2


paired_pb_sf_cd8 <- read_delim("~/working_dir/paired_pb_sf_cd8.txt", 
                               "\t", escape_double = FALSE, trim_ws = TRUE) %>% 
  filter(source == "PB") %>%
  select(status, donor, B27, B38, C12)



Patient <- data.frame(kmer=NaN, v=NaN, count.target=NaN, count.db=NaN, norm.count.db=NaN, norm.count.target=NaN, log2fc=NaN, p.binom=NaN, p.binom.adj=NaN, minus_log10p=NaN, name=NaN) 
Patient

for (i in name1){
  print("")
  print("_____________________________________________")
  print("")
  print(i)
  one=unlist(strsplit(i, split='_', fixed=TRUE))
  print(length(one))
  print(one)
  
  if (length(one) < 7){
    one=paste(one[-3], collapse="_")
    print(one)
  }
  if (length(one) == 7){
    one=paste(one[-c(3,5)], collapse="_")
    print(one)
  }
  
  
  for (j in name2){
    print(j)
    two=str_c(unlist(strsplit(j, split='_', fixed=TRUE)), sep = "_")
    print(length(two))
    print(two)
    
    if (length(two) < 7){
      two=paste(two[-3], collapse="_")
      print(two)
    }
    if (length(two) == 7){
      two=paste(two[-c(3,5)], collapse="_")
      print(two)
    }
    
    if(one == two){
      print("Found!")
      print(i)
      print(j)
      target = directory %+% j
      db = directory %+% i

      curr <- count_kmers(target = target, db = db, with.gaps = F,  target_preCalculated=T, db_precalculated=T)
      curr <- filter(curr, p.binom.adj < 0.01)
      data.sf <- fread(target)
      data.pb <- fread(db)
      name <- rep(unlist(strsplit(i, split='_', fixed=TRUE))[2], nrow(curr))
      curr["name"] = name
      Patient <- rbind(Patient, curr)
      next
    }
  }
  #break
}

colnames(Patient)[2] <- "V"
colnames(Patient)[11] <- "donor"
Patient <- na.omit(Patient)

total_table_B27_B38_C12 <- merge(Patient, paired_pb_sf_cd8, by = "donor") %>% 
  mutate(status = ifelse(status == "as", 1, 0)) %>%
  group_by(kmer, V) %>% 
  summarise(N=n(), N_B27=sum(B27), N_B38=sum(B38), N_C12=sum(C12), N_as=sum(status)) %>% 
  mutate(N_B27_neg = N - N_B27, N_B38_neg = N - N_B38, N_C12_neg = N - N_C12, N_psa = N - N_as)  %>%
  merge(Patient, by=c("kmer","V"))


```




```{r}
# Стартовые параметры
directory="/raid/users/pk/working_dir/mixcr_results_from_KK_CD8/"
files <- dir(directory)
files

`%+%` <- function(a, b) paste0(a, b)

file_names1 <- grep(".+_8.c.txt", files, value=TRUE)
file_names2 <- grep(".+_8_p..c.txt", files, value=TRUE)
file_names <- append(file_names1, file_names2)
file_names
print(length(file_names)/2)

# Собираю все последовательности из файлов
Patients_common_table <- data.frame(count=NaN, frequency=NaN, CDR3nt=NaN, CDR3aa=NaN, V=NaN, D=NaN, J=NaN, Vend=NaN, Dstart=NaN, Dend=NaN, Jstart=NaN, name=NaN) 
Patients_common_table
p = 0
for (i in file_names){
  print(i)
  p = p + 1
  curr <- read_delim(directory %+% "/" %+% i, "\t", escape_double = FALSE, trim_ws = TRUE)[,1:11]
  name <- rep( paste0(unlist(strsplit(i, split='_', fixed=TRUE))[1:3], collapse = "_"), nrow(curr))
  curr["name"] = name
  Patients_common_table <- rbind(Patients_common_table, curr)
}

Patients_common_table %<>% na.omit() %>% separate("name",into=c("condition","donor","source"))


K_mers_df <- data.frame(count=NaN, frequency=NaN, CDR3aa=NaN, CDR3nt=NaN, kmer=NaN, V=NaN, name_count=NaN, name=NaN)
K_mers_df

total <- total_table_B27_B38_C12
total <- filter(total, N > 2)


Patients_common_table <- Patients_common_table %>%  mutate(CDR3aa = str_sub(Patients_common_table$CDR3aa, 4, -2))



paired_pb_sf_cd8 <- read_delim("~/working_dir/paired_pb_sf_cd8.txt", "\t", escape_double = FALSE, trim_ws = TRUE)  %>%
  select(status, donor, source, B27, B38, C12)
paired_pb_sf_cd8 <- paired_pb_sf_cd8 %>%  mutate(name = apply(paired_pb_sf_cd8[1:3], 1, paste, collapse="_"))
```




```{r}

# Ищу последовательности CDR3 по к-мерам 
total.apply <- apply(X = total, MARGIN = 1,function(x){
  Patients_common_table %>%
    filter(V==x[2]) %>%
    filter(   grepl(x = Patients_common_table %>% 
                      filter(V==x[2]) %>% 
                      pull(CDR3aa), pattern=x[1])) %>% 
    mutate(kmer=x[1])
}) %>% rbindlist()  %>% mutate(len=nchar(CDR3aa))

total.apply %>% group_by(donor) %>% summarise(n=n()) #%>%  View()

# %>% separate(name, c("status", "donor", "source"), sep = "_")
k_mer_positions <- mapply(FUN=function(cdr3,kmer){
  regexpr(kmer,cdr3)[1]
}, total.apply$CDR3aa,total.apply$kmer) 

total.apply$k_mer_position = k_mer_positions

```


```{r}

# Разбиваем на два столбца столбец с частотами, получаем частоты по PB SF
PB_SF_freq_with_Na <- dcast(total.apply, CDR3nt+donor+V~source, value.var="frequency",fun.aggregate = mean)
PB_SF_freq_with_Na
# Заменяю наны на нули
PB_SF_freq <- replace(PB_SF_freq_with_Na, is.na(PB_SF_freq_with_Na), 0)
# Добавляю умеры, группирую по donor, kmer, V, суммирую частоты.
# Нашёл, с какой частотой встречается тот или иной к-мер в образцах в пациентах
total.apply_with_freq <- merge(PB_SF_freq, total.apply %>% select(-frequency, -donor, -V, -source), by="CDR3nt") %>%  group_by(donor, kmer, V) %>% summarise(sum_SF=sum(SF),sum_PB=sum(PB))

#  Ищу псевдокаунт для каждого образца как 0.1/UMI_sum
pseudocount <- Patients_common_table %>% filter(!is.nan(count)) %>% group_by(donor, source, condition) %>% summarise(UMI_sum=sum(count)) %>% mutate(pseudocount=0.1/UMI_sum) #%>% separate("name",into=c("condition","donor","source"))

#  Заменяю нули на псевдокаунты
total.apply_with_pseudocount <- merge(total.apply_with_freq, pseudocount %>% dcast(donor~source,value.var="pseudocount"),by="donor") %>% mutate(sum_SF=ifelse(sum_SF==0,SF,sum_SF),sum_PB=ifelse(sum_PB==0,PB,sum_PB) )

total.apply_with_pseudocount %<>% select(-SF,-PB) 
#  Ищу fold change
total.apply_with_pseudocount %<>% mutate(fold_change=sum_SF/sum_PB)
total.apply_with_pseudocount

# Группирую по кмеру и V гену
total.apply_with_fold_change <- total.apply_with_pseudocount %>% group_by(kmer,V) %>% summarise(fold_change_m=mean(fold_change))
total.apply_with_fold_change

```


```{r}

final_sure <- total.apply_with_fold_change %>% merge(unique(total %>% select(kmer,V,N_B27, N_B27_neg,N_B38,N_B38_neg,N,N_C12,N_C12_neg)), by=c("kmer","V"))

#write.csv(final_sure,directory %+% "final_sure.csv", row.names = FALSE)

```



```{r}

final_sure <- final_sure %>% mutate(Pr_B27 = N_B27/6)
final_sure <- final_sure %>% mutate(Pr_B27_neg = N_B27_neg/4)




# final_sure_plot4 %>% 
#   filter(V%in% HUMAN_TRBV) %>% 
#   filter(N_B27 > 3) %>%
#   ggplot(aes(x=fct_reorder(paste(kmer,V) ,fold_change_m),y=fold_change_m, fill = N_B27)) +
#     geom_histogram(stat="identity") +
#     theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
#   ggtitle('Fold change for 4-mers and V genes in B27+ donors') +
#   labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
#   scale_y_log10()


final_sure_plot4 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==4) %>% 
  filter(V%in% HUMAN_TRBV) %>% 
  filter(N_B27 > 3)


final_sure_plot4 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B27, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B27)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 4-mers and V genes in 6 B27+ donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')

final_sure_plot4 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B27_neg, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B27_neg)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 4-mers and V genes in 4 B27- donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')



final_sure_plot3 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==3) %>% 
  filter(V%in% HUMAN_TRBV) %>% 
  filter(N_B27 > 3)
  
  
  
final_sure_plot3 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B27, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B27)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 3-mers and V genes in 6 B27+ donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')

final_sure_plot3 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B27_neg, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B27_neg)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 3-mers and V genes in 4 B27- donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')

```


```{r}

final_sure <- final_sure %>% mutate(Pr_B38 = N_B38/3)
final_sure <- final_sure %>% mutate(Pr_B38_neg = N_B38_neg/7)



final_sure_plot4 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==4) %>% 
  filter(V%in% HUMAN_TRBV) %>% 
  filter(N_B38 > 1) %>% 
  filter(N_B38_neg > 3)


final_sure_plot4 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B38, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B38)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 4-mers and V genes in 3 B38+ donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')

final_sure_plot4 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B38_neg, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B38_neg)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 4-mers and V genes in 7 B38- donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')



final_sure_plot3 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==3) %>% 
  filter(V%in% HUMAN_TRBV) %>% 
  filter(N_B38 > 1) %>% 
  filter(N_B38_neg > 3)
  
  
  
final_sure_plot3 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B38, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B38)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 3-mers and V genes in 3 B38+ donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')


final_sure_plot3 %>% 
  ggplot(aes(x=fct_reorder2(fct_reorder(paste(kmer,V) ,fold_change_m), fold_change_m, Pr_B38_neg, .fun = last2, .desc = F), y=fold_change_m, fill = Pr_B38_neg)) +
    geom_histogram(stat="identity") +
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  ggtitle('Fold change for 3-mers and V genes in 7 B38- donors') +
  labs(x = "kmer + V gene", y = "fold change SF/PB in log10 scale") +
  scale_y_log10() +
  scale_fill_gradient(low ='light blue', high='navyblue')

```


```{r}

#######################################################################################################
final_sure <- final_sure %>% mutate(Pr_B27 = N_B27/6)
final_sure <- final_sure %>% mutate(Pr_B27_neg = N_B27_neg/4)


final_sure_plot4 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==4)

final_sure_plot3 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==3)


#lenKmer==4
ggplot(final_sure_plot4,aes(x=Pr_B27,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.1)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 4-меров и V генов B27+ доноров", 
       x = "% B27+ от 6 доноров", y = "fold change in log10 scale")+ geom_text(check_overlap = TRUE)
#lenKmer==3
ggplot(final_sure_plot3,aes(x=Pr_B27,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=1)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 3-меров и V генов B27+ доноров", 
       x = "% B27+ от 6 доноров", y = "fold change in log10 scale") + geom_text(check_overlap = TRUE)
#lenKmer==4
ggplot(final_sure_plot4,aes(x=Pr_B27_neg,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.4)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 4-меров и V генов B27- доноров", 
       x = "% B27- от 4 доноров", y = "fold change in log10 scale") + geom_text(check_overlap = TRUE)
#lenKmer==3
ggplot(final_sure_plot3,aes(x=Pr_B27_neg,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.8)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 3-меров и V генов B27- доноров", 
       x = "% B27- от 4 доноров", y = "fold change in log10 scale") + geom_text(check_overlap = TRUE)


```


```{r}
##############################################################

final_sure <- final_sure %>% mutate(Pr_B38 = N_B38/3)
final_sure <- final_sure %>% mutate(Pr_B38_neg = N_B38_neg/7)


final_sure_plot4 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==4)

final_sure_plot3 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==3)


#lenKmer==4
ggplot(final_sure_plot4,aes(x=Pr_B38,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.1)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 4-меров и V генов B38+ доноров", 
       x = "% B38+ от 3 доноров", y = "fold change in log10 scale")+ geom_text(check_overlap = TRUE)
#lenKmer==3
ggplot(final_sure_plot3,aes(x=Pr_B38,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=1)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 3-меров и V генов B38+ доноров", 
       x = "% B38+ от 3 доноров", y = "fold change in log10 scale")+ geom_text(check_overlap = TRUE)
#lenKmer==4
ggplot(final_sure_plot4,aes(x=Pr_B38_neg,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.4)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 4-меров и V генов B38- доноров", 
       x = "% B38+ от 3 доноров", y = "fold change in log10 scale")+ geom_text(check_overlap = TRUE)
#lenKmer==3
ggplot(final_sure_plot3,aes(x=Pr_B38_neg,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.8)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold change SF/PB для 3-меров и V генов B38- доноров", 
       x = "% B38+ от 3 доноров", y = "fold change in log10 scale")+ geom_text(check_overlap = TRUE)

```


```{r}
##########################################################

final_sure <- final_sure %>% mutate(Pr_C12 = N_C12/3)
final_sure <- final_sure %>% mutate(Pr_C12_neg = N_C12_neg/7)


final_sure_plot4 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==4)

final_sure_plot3 <- final_sure %>% mutate(lenKmer=nchar(kmer)) %>%  
  filter(lenKmer==3)


#lenKmer==4
ggplot(final_sure_plot4,aes(x=Pr_C12,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.1)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold_change_m PB и SF для 4-меров и V генов по C12+ донорам", 
       x = "% C12+ от 3 доноров", y = "fold_change_m")+ geom_text(check_overlap = TRUE)
#lenKmer==3
ggplot(final_sure_plot3,aes(x=Pr_C12,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=1)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold_change_m PB и SF для 3-меров и V генов по C12+ донорам", 
       x = "% C12+ от 3 доноров", y = "fold_change_m") + geom_text(check_overlap = TRUE)
#lenKmer==4
ggplot(final_sure_plot4,aes(x=Pr_C12_neg,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.4)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold_change_m PB и SF для 4-меров и V генов по C12- донорам", 
       x = "% C12- от 7 доноров", y = "fold_change_m") + geom_text(check_overlap = TRUE)
#lenKmer==3
ggplot(final_sure_plot3,aes(x=Pr_C12_neg,y=fold_change_m,label=paste(kmer,V))) +
  #geom_text(position=position_jitter(width=0, height=0.8)) +
  scale_y_log10() +
  geom_point() +
  labs(title = "Распределение fold_change_m PB и SF для 3-меров и V генов по C12- донорам", 
       x = "% C12- от 7 доноров", y = "fold_change_m") + geom_text(check_overlap = TRUE)


```


```{r}
#######################################################################################################

final_sure$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", final_sure$V)

g4 <- ggplot(final_sure, 
             aes(V_names, N)) 
g4 + stat_summary(fun.y = mean, geom = "bar",
                  color="blue", fill="white")

total.apply_with_freq %<>% mutate(sum_f = sum_SF+sum_PB) 
total.apply_with_freq$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", total.apply_with_freq$V)
total.apply$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", total.apply$V)


g1 <- ggplot(total.apply_with_freq,
             aes(y = sum_SF, x = sum_PB))

g1 + geom_point(aes(color = V, size = sum_f)) + 
  stat_smooth()+scale_y_log10()


V_summ <- total.apply_with_freq %>%  group_by(V) %>% summarise(allV_summ_V=n()/nrow(total.apply_with_freq)*100)




g7 <- ggplot(total.apply_with_freq,
             aes(x = V_names,
                 y = sum_f)) + 
  geom_point(aes(color = donor),
             alpha = 0.5,
             size = 2,
             position = position_jitter(width = 0.15, height = 0)) +scale_y_log10()
g7


g6 <- ggplot(total.apply_with_freq,
             aes(sum_PB, colour = V_names))
g6 + geom_freqpoly() 



ggplot(data=total.apply, aes(total.apply$len)) + 
  geom_histogram(aes(y =..density..), 
                 col="black", 
                 fill="blue", 
                 alpha=.30) + 
  scale_x_continuous(breaks=seq(0,30,1))



ggplot(total.apply, aes(x=as.factor(V_names),y=len))+geom_boxplot()+scale_y_log10()


g6 <- ggplot(total.apply,
             aes(len, colour = V))
g6 + geom_freqpoly() + 
  scale_x_continuous(breaks=seq(0,30,1))
g6
```

```{r}
ggplot(total.apply, aes(x=V,y=k_mer_position))+geom_boxplot()
```


```{r}
###################V usage по клонам###################################
df_PB <- total.apply %>% filter(source=="PB") %>% group_by(V) %>% summarise(N=n()/nrow(total.apply %>% filter(source=="PB"))*100)
df_PB$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", df_PB$V)

df_SF <- total.apply %>% filter(source=="SF") %>% group_by(V) %>% summarise(N=n()/nrow(total.apply %>% filter(source=="SF"))*100)
df <- df_PB %>% merge(df_SF, by="V")
df <- df %>% mutate(Vusage = log2(N.y/N.x))

ggplot(df,aes(V_names, Vusage)) + 
  stat_summary(fun.y = mean, geom = "bar",
                  color="blue", fill="white")+
  labs(title = "V usage по клонам ", 
       x = "V_names", y = "V usage")


```


```{r}
###################V usage по мясу###################################
df_PB <- total.apply %>% filter(source=="PB") %>% group_by(V, source) %>% summarise(N=sum(count)/sum(total.apply$count)*100)
df_PB$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", df_PB$V)

df_SF <- total.apply %>% filter(source=="SF") %>% group_by(V, source) %>% summarise(N=sum(count)/sum(total.apply$count)*100)
df <- df_PB %>% merge(df_SF, by="V")
df <- df %>% mutate(Vusage = log2(N.y/N.x))

ggplot(df,aes(V_names, Vusage)) + 
  stat_summary(fun.y = mean, geom = "bar",
               color="blue", fill="white")+
  labs(title = "V usage по мясу ", 
       x = "V_names", y = "V usage")




df2 <- total.apply %>% group_by(V, source, donor) %>% summarise(N=sum(count)/sum(total.apply$count))
df2$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", df2$V)

ggplot(df2,aes(V_names, N)) + 
  stat_summary(fun.y = mean, geom = "bar",
               color="blue", fill="white")+
  labs(title = "V usage по мясу ", 
       x = "V_names", y = "Доля V гена (%)")



# df2 <- total.apply  %>% group_by(V, source, donor) %>% summarise(N=sum(count)/sum(total.apply$count)*100)
# df2$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", df2$V)
# 
# df2$V_names = fct_reorder(df2$V_names, df2$N, max)


ggplot(data = df2, aes(x = V_names, y = N, fill = source)) + 
  labs(title = "V usage по клонам в PB и SF", 
       x = "V_names", y = "Доля V гена (%)") + 
  geom_histogram(stat="identity")+
  coord_flip()+
  facet_wrap(donor~.)+
  scale_y_log10()

df2$N <- order(df2$N, decreasing = FALSE)

ggplot(data = df2 %>% filter(donor== c("Ali", "Gor", "Shep")), aes(x = V_names, y = N, fill = source))+
  geom_bar(stat = "identity", position = position_dodge())+
  coord_flip()+
  facet_wrap(donor~.)+
  facet_wrap(~donor, scales = "free_x")


ggplot(data = df2, aes(x = V_names, y = N, fill = source))+
  geom_bar(stat = "identity", position = position_dodge())+
  coord_flip()+
  facet_wrap(donor~.)+
  facet_wrap(~donor, scales = "free_x")+
  labs(title = "Распределение fold_change PB и SF по V генам", 
       x = "V genes", y = "clone number")


# fc <- total.apply_with_pseudocount %>% group_by(V, donor) %>% summarise()




```


```{r}
#######################################################################################################
Patients_common_table_2 <- Patients_common_table %>% merge(paired_pb_sf_cd8 %>% select(donor, source, B27, B38, C12), by=c("source", "donor"))
Patients_common_table_2$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", Patients_common_table_2$V)

vusage_PB <- Patients_common_table_2 %>% filter(source=="PB" & B27=="1") %>% group_by(CDR3nt, V, donor) %>% summarise(nCLones=n(), nUMIs=sum(count))
vusage_SF <- Patients_common_table_2 %>% filter(source=="SF" & B27=="1") %>% group_by(CDR3nt, V, donor) %>% summarise(nCLones=n(), nUMIs=sum(count))
              
vusage <- merge(vusage_PB, vusage_SF, by=c("CDR3nt","V", "donor"), all=T) 
head(vusage)
vusage_cleaned <- vusage %>% mutate(nClones_PB=ifelse(is.na(nCLones.x)==T,min(nCLones.x,na.rm = T)/10,nCLones.x),nClones_SF=ifelse(is.na(nCLones.y)==T,min(nCLones.y,na.rm = T)/10,nCLones.y),nUMIs_SF=ifelse(is.na(nUMIs.x)==T,min(nUMIs.x,na.rm = T)/10,nUMIs.x),nUMIs_PB=ifelse(is.na(nUMIs.y)==T,min(nUMIs.y,na.rm = T)/10,nUMIs.y))
vusage_cleaned %<>% select(-nCLones.x, -nCLones.y, -nUMIs.x, -nUMIs.y) 

vusage_group <- vusage_cleaned %>% group_by(donor, V) %>% summarise(nClones_PB=sum(nClones_PB),nClones_SF=sum(nClones_SF),nUMIs_PB=sum(nUMIs_PB),nUMIs_SF=sum(nUMIs_SF))

vusage_fold <- vusage_group %>% mutate(fold_change_clones_log2 = log2(nClones_SF/nClones_PB)) %>% mutate(fold_change_UMI_log2 = log2(nUMIs_SF/nUMIs_PB))
head(vusage_fold)

vusage_fold$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", vusage_fold$V)

```


```{r}
#Clones
ggplot(data = vusage_fold, aes(x = V_names, y = fold_change_clones_log2))+
  geom_boxplot()+
  labs(title = "Распределение fold_change по клонам PB и SF для V генов C12+", 
       x = "V genes", y = "fold change")


#Clones
theme_set(theme_classic())
a <- ggplot(vusage_fold, aes(x = V_names, y = fold_change_clones_log2))+
  labs(title = "Fold change по клонам PB/SF для V генов B27+", 
       x = "V genes", y = "log 10 PB/SF")+ 
  geom_boxplot(outlier.colour = "blue", outlier.alpha = 0.2, outlier.size = 3, outlier.shape = 21, outlier.stroke = 2)+
  geom_hline(yintercept=0, linetype="dashed", color = "black")+
  geom_point(position=position_jitter(0.2), size = 3, shape = 21, color = "blue", alpha = 0.2, stroke = 2) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

b <-ggplot(Patients_common_table_2 %>% filter(B27=="1"),aes(x = V_names, fill=count)) + 
  labs(title = "Распределение плотности по V генам B27+", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black", fill="#FFFFFF") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)

```
```{r}
#UMI
ggplot(data = vusage_fold, aes(x = V_names, y = fold_change_UMI_log2))+
  geom_boxplot()+
  labs(title = "Распределение fold_change по UMI PB и SF для V генов B27-", 
       x = "V genes", y = "fold change")


```



```{r}

#######################################################################################################
Patients_common_table_2 <- Patients_common_table %>% merge(paired_pb_sf_cd8 %>% select(donor, source, B27, B38, C12), by=c("source", "donor"))

vusage_PB <- Patients_common_table_2 %>% filter(source=="PB" & B27=="1") %>% group_by(CDR3nt, V, donor) %>% summarise(nCLones=n(), nUMIs=sum(count))
vusage_SF <- Patients_common_table_2 %>% filter(source=="SF" & B27=="1") %>% group_by(CDR3nt, V, donor) %>% summarise(nCLones=n(), nUMIs=sum(count))

vusage <- merge(vusage_PB, vusage_SF, by=c("CDR3nt","V", "donor"), all=T)
# head(vusage)
vusage_cleaned <- vusage %>% mutate(nClones_PB=ifelse(is.na(nCLones.x)==T,min(nCLones.x,na.rm = T)/10,nCLones.x),nClones_SF=ifelse(is.na(nCLones.y)==T,min(nCLones.y,na.rm = T)/10,nCLones.y),nUMIs_SF=ifelse(is.na(nUMIs.x)==T,min(nUMIs.x,na.rm = T)/10,nUMIs.x),nUMIs_PB=ifelse(is.na(nUMIs.y)==T,min(nUMIs.y,na.rm = T)/10,nUMIs.y))
vusage_cleaned %<>% select(-nCLones.x, -nCLones.y, -nUMIs.x, -nUMIs.y)

vusage_group <- vusage_cleaned %>% group_by(donor, V) %>% summarise(nClones_PB=sum(nClones_PB),nClones_SF=sum(nClones_SF),nUMIs_PB=sum(nUMIs_PB),nUMIs_SF=sum(nUMIs_SF))

vusage_fold <- vusage_group %>% mutate(fold_change_clones_log10 = log10(nClones_PB/nClones_SF)) %>% mutate(fold_change_UMI_log10 = log10(nUMIs_PB/nUMIs_SF))
# head(vusage_fold)

vusage_fold$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", vusage_fold$V)
Patients_common_table_2$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", Patients_common_table_2$V)


```


```{r}
#Clones

a <- ggboxplot(data = vusage_fold, x = "V_names", y = "fold_change_clones_log10", sort.by.groups = TRUE)+
  labs(title = "Распределение fold_change и плотность по клонам PB/SF для V генов B27+", 
       x = "V genes", y = "log 10 PB/SF")+ 
  geom_boxplot(outlier.colour = "blue", outlier.alpha = 0.4, outlier.size = 3, outlier.shape = 21, outlier.stroke = 2)+
  geom_hline(yintercept=0, linetype="dashed", color = "black")+
  geom_point(position=position_jitter(0.2), size = 3, shape = 21, color = "blue", alpha = 0.4, stroke = 2) +
  theme_classic()+ 
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

b <-ggplot(Patients_common_table_2 %>% filter(B38=="1"),aes(x=V_names, fill=count)) + geom_bar(alpha=.8) + theme(axis.text.x = element_text(angle = 90, hjust = 1)) + guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)




```



```{r}
#######################################################################################################
Patients_common_table_2 <- Patients_common_table %>% merge(paired_pb_sf_cd8 %>% select(donor, source, B27, B38, C12), by=c("source", "donor"))

vusage_PB <- Patients_common_table_2 %>% filter(source=="PB" & B27=="1") %>% group_by(V, donor) %>% summarise(nCLones=n(), nUMIs=sum(count))
vusage_SF <- Patients_common_table_2 %>% filter(source=="SF" & B27=="1") %>% group_by(V, donor) %>% summarise(nCLones=n(), nUMIs=sum(count))

vusage <- merge(vusage_PB, vusage_SF, by=c("V", "donor"), all=T) 
# head(vusage)
vusage_cleaned <- vusage %>% mutate(nClones_PB=ifelse(is.na(nCLones.x)==T,min(nCLones.x,na.rm = T)/10,nCLones.x),nClones_SF=ifelse(is.na(nCLones.y)==T,min(nCLones.y,na.rm = T)/10,nCLones.y),nUMIs_SF=ifelse(is.na(nUMIs.x)==T,min(nUMIs.x,na.rm = T)/10,nUMIs.x),nUMIs_PB=ifelse(is.na(nUMIs.y)==T,min(nUMIs.y,na.rm = T)/10,nUMIs.y))
vusage_cleaned %<>% select(-nCLones.x, -nCLones.y, -nUMIs.x, -nUMIs.y) 

vusage_group <- vusage_cleaned %>% group_by(donor, V) %>% summarise(nClones_PB=sum(nClones_PB),nClones_SF=sum(nClones_SF),nUMIs_PB=sum(nUMIs_PB),nUMIs_SF=sum(nUMIs_SF))

vusage_fold <- vusage_group %>% mutate(fold_change_clones_log10 = log10(nClones_PB/nClones_SF)) %>% mutate(fold_change_UMI_log10 = log10(nUMIs_PB/nUMIs_SF))
# head(vusage_fold)

vusage_fold$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", vusage_fold$V)
Patients_common_table_2$V_names <- gsub("TRBV([0-9]?.+).*", "\\1", Patients_common_table_2$V)



```


```{r}
#Clones
theme_set(theme_classic())
a <- ggplot(vusage_fold, aes(x = V_names, y = fold_change_clones_log10))+
  labs(title = "Fold change по клонам PB/SF для V генов B27+", 
       x = "V genes", y = "log 10 PB/SF")+ 
  geom_boxplot(outlier.colour = "blue", outlier.alpha = 0.2, outlier.size = 3, outlier.shape = 21, outlier.stroke = 2)+
  geom_hline(yintercept=0.5, linetype="dashed", color = "black")+
  geom_point(position=position_jitter(0.2), size = 3, shape = 21, color = "blue", alpha = 0.2, stroke = 2) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

b <-ggplot(Patients_common_table_2 %>% filter(B27=="1"),aes(x=V_names, fill=count)) + 
  labs(title = "Распределение плотности по V генам B27+", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black", fill="#FFFFFF") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)


```


```{r}
corr_fc <- apply(X = vusage_fold, MARGIN = 1,function(x){
  tmp <- Patients_common_table_2 %>%
    filter(V==x[2]) %>%
    group_by(V) %>% 
    summarise(N=n()) %>% .$N
  as.numeric(x[7])/tmp
})
corr_fc

vusage_fold$corr_fc <- corr_fc


#Clones
theme_set(theme_classic())
a <- ggplot(vusage_fold, aes(x = V_names, y = corr_fc))+
  labs(title = "Fold change по клонам PB/SF для V генов B27+", 
       x = "V genes", y = "log 10 PB/SF")+ 
  geom_boxplot(outlier.colour = "blue", outlier.alpha = 0.2, outlier.size = 3, outlier.shape = 21, outlier.stroke = 2)+
  geom_hline(yintercept=0, linetype="dashed", color = "black")+
  geom_point(position=position_jitter(0.2), size = 3, shape = 21, color = "blue", alpha = 0.2, stroke = 2) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))

b <-ggplot(Patients_common_table_2 %>% filter(B27=="1"),aes(x=V_names, fill=count)) + 
  labs(title = "Распределение плотности по V генам B27+", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black", fill="#FFFFFF") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)


```


```{r}

################################pbinom####################################

vusage_cleaned %<>% mutate(nClones_PB=as.integer(nClones_PB), nClones_SF=as.integer(nClones_SF)) 
vusage_cleaned <- vusage_cleaned %>% 
  mutate(pbinom= pbinom(vusage_cleaned$nClones_SF, sum(vusage_cleaned$nClones_SF), 
                       vusage_cleaned$nClones_PB/sum(vusage_cleaned$nClones_PB), 
                       lower.tail = F))

vusage_cleaned %>% group_by(V) %>% mutate(p_sum = sum(pbinom))

Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  group_by(V,donor,source) %>%
  summarise(Ncl=n()) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>% 
  group_by(donor) %>% 
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>% 
  filter(PB!=0 & SF!=0) %>% 
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 
  merge(vusage_cleaned %>% select(V, donor, pbinom), by=c("V", "donor")) %>% 
  group_by(V) %>% mutate(p_sum = round(mean(pbinom),4)) %>% 
  
  ggplot(aes(x=V_names, y=logfc, group=V_names)) +
  ggtitle('V usage by cln PB to SF CD8 B27+') +
  geom_point(aes(color=pbinom<0.05, fill=pbinom<0.05),alpha=0.8, shape=21, size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed') #+
  #geom_text(aes(label = round(pbinom,4)), check_overlap = TRUE)
  
```


```{r}
# Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV) %>% 
#   group_by(V,donor,source) %>%
#   summarise(Ncl=n()) %>%
#   dcast(V+donor~source,value.var = "Ncl") %>%
#   replace_na(list(PB=0,SF=0)) %>% 
#   group_by(donor) %>% 
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
#   ungroup() %>% 
#   filter(PB!=0 & SF!=0) %>% 
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 
# 
#   ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names)) +
#   ggtitle('V usage by clone number PB to SF CD8 B27+') +
#   geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
#   geom_boxplot(alpha=0)+ theme_classic() +
#   theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
#   scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
#   geom_hline(yintercept = 1, linetype='dashed')




#Clones B27 мясо 
theme_set(theme_classic())
a <- Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  filter(B27=="1") %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=n()) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>% 
  group_by(donor) %>% 
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>% 
  filter(PB!=0 & SF!=0) %>% 
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
  ggtitle('V usage by clone number PB to SF CD8 B27+') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')+
   guides(fill=FALSE)

b <-ggplot(Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
    filter(B27=="1") %>% 
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) + 
  labs(title = "Density distribution by V genes in B27+", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)

#Clones B38
theme_set(theme_classic())
a <- Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  filter(B38=="1") %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=n()) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>% 
  group_by(donor) %>% 
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>% 
  filter(PB!=0 & SF!=0) %>% 
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
  ggtitle('V usage by clone number PB to SF CD8 B38+') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')+
   guides(fill=FALSE)

b <-ggplot(Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
    filter(B38=="1") %>% 
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) + 
  labs(title = "Density distribution by V genes in B38+", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)

#Clones
theme_set(theme_classic())
a <- Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  group_by(V,donor,source) %>%
  summarise(Ncl=n()) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>% 
  group_by(donor) %>% 
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>% 
  filter(PB!=0 & SF!=0) %>% 
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
  ggtitle('V usage by clone number PB to SF CD8 All') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')+
   guides(fill=FALSE)

b <-ggplot(Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) + 
  labs(title = "Density distribution by V genes in All", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)


b
```

```{r}

#Clones клоны
# theme_set(theme_classic())
# a <- Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV) %>%
#   group_by(V,donor,source) %>%
#   summarise(Ncl=n()) %>%
#   dcast(V+donor~source,value.var = "Ncl") %>%
#   replace_na(list(PB=0,SF=0)) %>%
#   group_by(donor) %>%
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
#   ungroup() %>%
#   filter(PB!=0 & SF!=0) %>%
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>%
# 
#   ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
#   ggtitle('V usage by clone number PB to SF CD8 B27+') +
#   geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
#   geom_boxplot(alpha=0)+ theme_classic() +
#   theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
#   scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
#   geom_hline(yintercept = 1, linetype='dashed')+
#   guides(fill=FALSE)
# 
# b <-ggplot(Patients_common_table_2 %>%
#     filter(V%in% HUMAN_TRBV) %>%
#     group_by(V,donor,source,B27) %>%
#     summarise(Ncl=n()) %>%
#     mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>%
#     filter(B27=="1"),aes(x=V_names, fill=Ncl)) +
#     labs(title = "Density of distribution by V genes in B27+",
#        x = "V genes", y = "clone number")+
#     geom_bar(alpha=.8, colour="black", fill="#FFFFFF") +
#     theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
#     guides(fill=FALSE)
# 
# ggarrange(a, b,
#           labels = c("A", "B"),
#           ncol = 1, nrow = 2)

# Patients_common_table_2 %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V))
# B27_Patients_common_table_2 <- Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV) %>% 
#   filter(B27=="1") 
# colnames(B27_Patients_common_table_2)[1] = "Source"
# 
# #Clones B27
# theme_set(theme_classic())
# a <- B27_Patients_common_table_2 %>%
#   group_by(V,donor,Source) %>%
#   summarise(count=sum(count)) %>%
#   dcast(V+donor~Source,value.var = "count") %>%
#   replace_na(list(PB=0,SF=0)) %>% 
#   group_by(donor) %>% 
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>%
#   ungroup() %>% 
#   filter(PB!=0 & SF!=0) %>% 
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 
#   ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
#   ggtitle('V usage by clone count PB to SF CD8 B27+') +
#   geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
#   geom_boxplot(alpha=0)+ theme_classic() +
#   theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
#   scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
#   geom_hline(yintercept = 1, linetype='dashed') +
#    guides(fill=FALSE)
# 
# b <- B27_Patients_common_table_2 %>% 
#   group_by(V) %>% 
#   summarise(count = sum(count)) %>% 
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 
#   ggplot(aes(x=V_names, y=Source)) +
#   geom_bar(stat="identity", colour="black") +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
#   guides(fill=FALSE)
# 
# ggarrange(a, b, 
#           labels = c("A", "B"),
#           ncol = 1, nrow = 2)
# 
# 
# B38_Patients_common_table_2 <- Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV) %>% 
#     filter(B38=="1") 
# 
# #Clones B27
# theme_set(theme_classic())
# a <- B38_Patients_common_table_2 %>%
#   group_by(V,donor,source) %>%
#   summarise(count=sum(count)) %>%
#   dcast(V+donor~source,value.var = "count") %>%
#   replace_na(list(PB=0,SF=0)) %>% 
#   group_by(donor) %>% 
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
#   ungroup() %>% 
#   filter(PB!=0 & SF!=0) %>% 
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 
# 
#   ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
#   ggtitle('V usage by clone count PB to SF CD8 B38+') +
#   geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
#   geom_boxplot(alpha=0)+ theme_classic() +
#   theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
#   scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
#   geom_hline(yintercept = 1, linetype='dashed')+
#    guides(fill=FALSE)
# 
# b <- ggplot(B38_Patients_common_table_2 %>% group_by(V) %>% summarise(count = sum(count)) %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)), aes(x=V_names, y=source)) +
#   geom_bar(stat="identity", colour="black")  +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
#   guides(fill=FALSE)
# 
# ggarrange(a, b, 
#           labels = c("A", "B"),
#           ncol = 1, nrow = 2)
# 
# All_Patients_common_table_2 <- Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV)
# 
# #Clones B27
# theme_set(theme_classic())
# a <- All_Patients_common_table_2 %>%
#   group_by(V,donor,source) %>%
#   summarise(count=sum(count)) %>%
#   dcast(V+donor~source,value.var = "count") %>%
#   replace_na(list(PB=0,SF=0)) %>% 
#   group_by(donor) %>% 
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
#   ungroup() %>% 
#   filter(PB!=0 & SF!=0) %>% 
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% #group_by(donor) %>%  summarise() %>%  View()
# 
#   ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names), sort.by.groups = TRUE) +
#   ggtitle('V usage by clone count PB to SF CD8 All') +
#   geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
#   geom_boxplot(alpha=0)+ theme_classic() +
#   theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1)) +
#   scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
#   geom_hline(yintercept = 1, linetype='dashed')+
#    guides(fill=FALSE)
# 
# b <- ggplot(All_Patients_common_table_2 %>% group_by(V) %>% summarise(count = sum(count)) %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)), aes(x=V_names, y=source)) +
#   geom_bar(stat="identity", colour="black")  +
#   theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
#   guides(fill=FALSE)
# 
# ggarrange(a, b, 
#           labels = c("A", "B"),
#           ncol = 1, nrow = 2)

#/sum(B27_Patients_common_table_2$count)*100
```



```{r}
Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
  group_by(donor,source) %>%
  mutate(frequency_count = count/sum(count)) %>%
  ungroup() %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=sum(frequency_count)) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>%
  group_by(donor) %>%
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>%
  filter(PB!=0 & SF!=0) %>%
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>%

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names)) +
  ggtitle('V usage by clone count PB to SF CD8 B27+') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')


#Clones
theme_set(theme_classic())
a <-Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
  filter(B27==1) %>%
  group_by(donor,source) %>%
  mutate(frequency_count = count/sum(count)) %>%
  ungroup() %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=sum(frequency_count)) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>%
  group_by(donor) %>%
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>%
  filter(PB!=0 & SF!=0) %>%
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% #%>% summarize("donor") %>% View()

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names)) +
  ggtitle('V usage by clone count PB to SF CD8 B27+') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')
  guides(fill=FALSE)

b <-ggplot(Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
    filter(B27=="1") %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) +
  labs(title = "Density of distribution by V genes in B27+",
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  guides(fill=FALSE)

ggarrange(a, b,
          labels = c("A", "B"),
          ncol = 1, nrow = 2)



#Clones
theme_set(theme_classic())
a <-Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
  filter(B38==1) %>%
  group_by(donor,source) %>%
  mutate(frequency_count = count/sum(count)) %>%
  ungroup() %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=sum(frequency_count)) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>%
  group_by(donor) %>%
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>%
  filter(PB!=0 & SF!=0) %>%
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% #%>% summarize("donor") %>% View()

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names)) +
  ggtitle('V usage by clone count PB to SF CD8 B38+') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')
  guides(fill=FALSE)

b <-ggplot(Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
    filter(B38=="1") %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) +
  labs(title = "Density of distribution by V genes in B38+",
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  guides(fill=FALSE)

ggarrange(a, b,
          labels = c("A", "B"),
          ncol = 1, nrow = 2)


#Clones
theme_set(theme_classic())
a <-Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
  filter(C12==1) %>%
  group_by(donor,source) %>%
  mutate(frequency_count = count/sum(count)) %>%
  ungroup() %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=sum(frequency_count)) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>%
  group_by(donor) %>%
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>%
  filter(PB!=0 & SF!=0) %>%
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% #%>% summarize("donor") %>% View()

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names)) +
  ggtitle('V usage by clone count PB to SF CD8 C12+') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')
  guides(fill=FALSE)

b <-ggplot(Patients_common_table_2 %>%
  filter(V %in% HUMAN_TRBV) %>%
    filter(C12=="1") %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) +
  labs(title = "Density of distribution by V genes in C12+",
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  guides(fill=FALSE)

  
ggarrange(a, b,
          labels = c("A", "B"),
          ncol = 1, nrow = 2)

#Clones
theme_set(theme_classic())
a <-Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>%
  group_by(donor,source) %>%
  mutate(frequency_count = count/sum(count)) %>%
  ungroup() %>%
  group_by(V,donor,source) %>%
  summarise(Ncl=sum(frequency_count)) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>%
  group_by(donor) %>%
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>%
  filter(PB!=0 & SF!=0) %>%
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% #%>% summarize("donor") %>% View()

  ggplot(aes(x=V_names, y=logfc, fill=logfc>1, group=V_names)) +
  ggtitle('V usage by clone count PB to SF CD8 All') +
  geom_point(alpha=0.8, shape=21, color='black', size=2.5)  +
  geom_boxplot(alpha=0)+ theme_classic() +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  scale_y_log10() + xlab('') + ylab('log10 PB/SF cln') +
  geom_hline(yintercept = 1, linetype='dashed')
  guides(fill=FALSE)


b <-ggplot(Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)),aes(x=V_names, fill=source)) +
  labs(title = "Density of distribution by V genes in All+",
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black") +
  theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1, size=20), axis.text.y = element_text(vjust=0.5, hjust = 1,  size=20), 
        axis.title = element_text(size = 20), plot.title = element_text(size=25)) +
  guides(fill=FALSE)

ggarrange(a, b,
          labels = c("A", "B"),
          ncol = 1, nrow = 2)
```

```{r}
################################pbinom####################################

vusage_cleaned %<>% mutate(nClones_PB=as.integer(nClones_PB), nClones_SF=as.integer(nClones_SF)) 
vusage_cleaned <- vusage_cleaned %>% 
  mutate(pbinom= pbinom(vusage_cleaned$nClones_SF, sum(vusage_cleaned$nClones_SF), 
                        vusage_cleaned$nClones_PB/sum(vusage_cleaned$nClones_PB), 
                        lower.tail = F)) %>% mutate(pbinom_adj = p.adjust(pbinom))

vusage_cleaned %>% group_by(V) %>% mutate(p_sum = sum(pbinom_adj))

library(forcats)

# Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV) %>% 
#   group_by(V,donor,source) %>%
#   summarise(Ncl=n()) %>%
#   dcast(V+donor~source,value.var = "Ncl") %>%
#   replace_na(list(PB=0,SF=0)) %>% 
#   group_by(donor) %>% 
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
#   ungroup() %>% 
#   filter(PB!=0 & SF!=0) %>% 
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>% 
#   merge(vusage_cleaned %>% select(V, donor, pbinom), by=c("V", "donor")) %>% 
#   group_by(V) %>% mutate(p_sum = round(mean(pbinom),4)) %>% 
#   dcast(V~donor,value.var = "pbinom") %>% 
#   mutate(N_enriched=rowSums(.[2:ncol(.)]<0.05)   ) %>%
#   mutate(N_enriched=N_enriched/(ncol(.)-2)) %>% na.omit() %>% 
#   ggplot(aes(x=fct_reorder(V,N_enriched),y=N_enriched))+
#     geom_histogram(stat="identity")+
#     theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1))+
#   ggtitle('Number of enriched V genes by clone number in % of 8 donors') +
#   geom_hline(yintercept = 1, linetype='dashed') +
#   labs(x = "V genes", y = "N of donors enriched")
  


t_p <- Patients_common_table_2 %>%
  filter(V %in% HUMAN_TRBV) %>% 
  filter(B27=="1") %>% 
  group_by(V,donor,source) %>%
  summarise(Ncl=n()) %>%
  dcast(V+donor~source,value.var = "Ncl") %>%
  replace_na(list(PB=0,SF=0)) %>% 
  group_by(donor) %>% 
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>% 
  filter(PB!=0 & SF!=0) %>% 
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  merge(vusage_cleaned %>% 
  select(V, donor, pbinom_adj), by=c("V", "donor")) %>% #group_by(donor) %>% summarise() %>%  View()
  group_by(V) %>% mutate(p_sum = round(mean(pbinom_adj),4)) %>% 
  dcast(V~donor,value.var = "pbinom_adj") %>% 
  mutate(N_enriched=rowSums(.[2:ncol(.)]<0.05)   ) %>%
  mutate(N_enriched=N_enriched/(ncol(.)-2)*100) %>% 
  na.omit() %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V))# %>% #View()

t_p2 <- Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  filter(B27=="1") %>% 
  merge(t_p, by="V") %>% 
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V))

  
#Clones
theme_set(theme_classic())
a <- t_p %>% 
  ggplot(aes(x=fct_reorder(V_names,N_enriched),y=N_enriched))+
    geom_histogram(stat="identity")+
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1))+
  ggtitle('V usage by clone number PB/SF CD8 B27+ in % of 8 donors') +
  geom_hline(yintercept = 1, linetype='dashed') +
  labs(x = "V genes", y = "N of donors enriched")

b <- t_p2 %>% 
  ggplot(aes(x=fct_reorder(V_names,N_enriched), fill=count)) + 
  labs(title = "Density of distribution by V genes in B27+", 
       x = "V genes", y = "density")+
  geom_bar(alpha=.8, colour="black", fill="#FFFFFF") + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
  guides(fill=FALSE)


ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)


```
```{r}
################################pbinom####################################

# vusage_cleaned %<>% mutate(nClones_PB=as.integer(nClones_PB), nClones_SF=as.integer(nClones_SF)) 
# vusage_cleaned <- vusage_cleaned %>% 
#   mutate(pbinom= pbinom(vusage_cleaned$nClones_SF, sum(vusage_cleaned$nClones_SF), 
#                         vusage_cleaned$nClones_PB/sum(vusage_cleaned$nClones_PB), 
#                         lower.tail = F)) %>% mutate(pbinom_adj = p.adjust(pbinom))

# vusage_cleaned %>% group_by(V) %>% mutate(p_sum = sum(pbinom_adj))

# library(forcats)

# Patients_common_table_2 %>%
#   filter(V%in% HUMAN_TRBV) %>%
#   group_by(V,donor,source) %>%
#   summarise(count=sum(count)) %>%
#   dcast(V+donor~source,value.var = "count") %>%
#   replace_na(list(PB=0,SF=0)) %>%
#   group_by(donor) %>%
#   mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
#   ungroup() %>%
#   filter(PB!=0 & SF!=0) %>%
#   mutate(logfc=PBnorm/SFnorm) %>%
#   group_by(donor) %>%
#   mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V)) %>%
#   merge(vusage_cleaned %>% select(V, donor, pbinom_adj), by=c("V", "donor")) %>%
#   group_by(V) %>% mutate(p_sum = round(mean(pbinom_adj),4)) %>%
#   dcast(V~donor,value.var = "pbinom_adj") %>%
#   mutate(N_enriched=rowSums(.[2:ncol(.)]<0.05)   ) %>%
#   mutate(N_enriched=N_enriched/(ncol(.)-2)) %>% na.omit() %>%
#   ggplot(aes(x=fct_reorder(V,N_enriched),y=N_enriched))+
#     geom_histogram(stat="identity")+
#     theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1))+
#   ggtitle('Number of enriched V genes by clone count in % of 8 donors') +
#   geom_hline(yintercept = 1, linetype='dashed') +
#   labs(x = "V genes", y = "N of donors enriched")



t_p <- Patients_common_table_2 %>%
  filter(V %in% HUMAN_TRBV) %>% 
  filter(B27=="1") %>%
  group_by(V,donor,source) %>%
  summarise(count=sum(count)) %>%
  dcast(V+donor~source,value.var = "count") %>%
  replace_na(list(PB=0,SF=0)) %>% 
  group_by(donor) %>% 
  mutate(PBnorm=PB/sum(PB),SFnorm=SF/sum(SF)) %>% #View (title="ours")
  ungroup() %>% 
  filter(PB!=0 & SF!=0) %>% 
  mutate(logfc=PBnorm/SFnorm) %>%
  group_by(donor) %>%
  merge(vusage_cleaned %>% 
  select(V, donor, pbinom_adj), by=c("V", "donor")) %>% #group_by(donor) %>% summarise() %>%  View()
  group_by(V) %>% mutate(p_sum = round(mean(V),4)) %>% 
  dcast(V~donor,value.var = "pbinom_adj") %>% 
  mutate(N_enriched=rowSums(.[2:ncol(.)]<0.05)   ) %>%
  mutate(N_enriched=N_enriched/(ncol(.)-2)*100) %>% 
  na.omit() %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V))# %>% #View()

t_p2 <- Patients_common_table_2 %>%
  filter(V%in% HUMAN_TRBV) %>% 
  filter(B27=="1") %>% 
  group_by(V) %>% summarise(count = sum(count)) %>%
  merge(t_p, by="V") %>%
  mutate(V_names=gsub("TRBV([0-9]?.+).*", "\\1", V))# %>% View()

  
#Clones
theme_set(theme_classic())
a <- t_p %>% 
  ggplot(aes(x=fct_reorder(V_names,N_enriched),y=N_enriched))+
    geom_histogram(stat="identity")+
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1))+
  ggtitle('V usage by clone count PB/SF CD8 B27+ in % of 8 donors') +
  geom_hline(yintercept = 1, linetype='dashed') +
  labs(x = "V genes", y = "N of donors enriched")

b <- t_p2 %>%   ggplot(aes(x=fct_reorder(V_names,N_enriched),y=count))+
    labs(title = "Density of distribution by V genes in B27+", 
       x = "V genes", y = "density")+
    geom_histogram(stat="identity",alpha=.8, colour="black", fill="#FFFFFF")+
    theme(axis.text.x = element_text(angle = 90, vjust=0.5, hjust = 1))+
    theme(axis.text.x = element_text(angle = 90, hjust = 1)) + 
    guides(fill=FALSE)

ggarrange(a, b, 
          labels = c("A", "B"),
          ncol = 1, nrow = 2)


```


```{r}

```



```{r}
library(data.table)

paired_pb_sf_cd8 <- read_delim("~/working_dir/paired_pb_sf_cd8.txt", "\t", escape_double = FALSE, trim_ws = TRUE)

Patients_common_table_2 <- Patients_common_table %>% merge(paired_pb_sf_cd8 %>% select(donor, source, B27, B38, C12), by=c("source", "donor"))
#View(Patients_common_table_2)

tab <- Patients_common_table_2 %>% 
  dcast(V+donor+CDR3nt+CDR3aa+B27+B38+C12+condition~source,value.var = "frequency", fill = 1, fun.aggregate = sum)
tab

tab <- tab %>% group_by(donor) %>% mutate(PB=ifelse(PB==1,min(PB)/10,PB), SF=ifelse(SF==1,min(SF)/10,SF))
#View(tab)


selected_petterns <- rbind(tab[grep("SIRS", tab$CDR3aa), ] %>% filter(V=="TRBV19") %>% mutate(pattern_name = "SIRS"), tab[grep("YSTD", tab$CDR3aa), ]%>% filter(V=="TRBV9") %>% mutate(pattern_name = "YSTD"), tab[grep("GANV", tab$CDR3aa), ] %>% filter(V=="TRBV4-3") %>% mutate(pattern_name = "GANV"), tab[grep("GDTG", tab$CDR3aa), ] %>% filter(V=="TRBV20-1") %>% mutate(pattern_name = "GDTG"),
        tab[grep("DYNE", tab$CDR3aa), ] %>% filter(V=="TRBV20-1") %>% mutate(pattern_name = "DYNE"),
        tab[grep("DTGE", tab$CDR3aa), ] %>% filter(V=="TRBV20-1") %>% mutate(pattern_name = "DTGE")) 

# View(selected_petterns)

selected_petterns <- selected_petterns %>%
  mutate(MHC_status = case_when(B27 == 1 ~ "B27", C12 == 1 ~ "B38/C12", B38 == 1 ~ "B38/C12")) %>% mutate(fc=PB/SF)
#View(selected_petterns)


```


```{r}
ggplot(selected_petterns, aes(x=pattern_name, y=fc, fill = condition)) +
  ggtitle('Fold change by clone count CD8 in all as and psa') +
  geom_boxplot(alpha=0.5) +
  ylab('PB/SF for CDR3aa + V gene in log10 scale')  + scale_y_log10()+
  theme_classic()+
  geom_hline(yintercept = 1, linetype='dashed')+
  geom_point(aes(fill = condition), size = 5, shape = 21, alpha=0.8, position = position_jitterdodge()) +
  theme(text = element_text(size = 18),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

ggplot(selected_petterns %>% filter(B38 == 1), aes(x=pattern_name, y=fc, fill = condition)) +
  ggtitle('Fold change by clone count CD8 in B38+ as and psa') +
  geom_boxplot(alpha=0.5) +
  ylab('PB/SF for CDR3aa + V gene in log10 scale')  + scale_y_log10()+
  theme_classic()+
  geom_hline(yintercept = 1, linetype='dashed')+
  geom_point(aes(fill = condition), size = 5, shape = 21, alpha=0.8, position = position_jitterdodge()) +
  theme(text = element_text(size = 18),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

ggplot(selected_petterns %>% filter(B27 == 1), aes(x=pattern_name, y=fc, fill = condition)) +
  ggtitle('Fold change by clone count CD8 in B27+ as and psa') +
  geom_boxplot(alpha=0.5) +
  ylab('PB/SF for CDR3aa + V gene in log10 scale')  + scale_y_log10()+
  theme_classic()+
  geom_hline(yintercept = 1, linetype='dashed')+
  geom_point(aes(fill = condition), size = 5, shape = 21, alpha=0.8, position = position_jitterdodge()) +
  theme(text = element_text(size = 18),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())

```

```{r}
ggplot(selected_petterns, aes(x=pattern_name, y=fc, fill = MHC_status)) +
  ggtitle('Fold change by clone count CD8 in all donors') +
  geom_boxplot(alpha=0.5) +
  ylab('PB/SF for CDR3aa + V gene in log10 scale')  + scale_y_log10()+
  theme_classic()+
  geom_hline(yintercept = 1, linetype='dashed')+
  geom_point(aes(fill = MHC_status), size = 5, shape = 21, alpha=0.8, position = position_jitterdodge()) +
  theme(text = element_text(size = 18),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        panel.grid.minor.x = element_blank(),
        panel.grid.major.x = element_blank())
```














```{r}
#######################################################################################################
# tab <- total.apply %>% 
#   dcast(count+V+donor+CDR3nt+CDR3aa+kmer+condition~source,value.var = "frequency", fill = 1, fun.aggregate = sum)
# View(tab)
# 
# tab <- tab %>% group_by(donor) %>% mutate(PB=ifelse(PB==1,min(PB)/10,PB), SF=ifelse(SF==1,min(SF)/10,SF))
# View(tab)
# 
# matrix_table <- tab %>% 
#   mutate(pbinom= pbinom(vusage_cleaned$nClones_SF, sum(vusage_cleaned$nClones_SF), 
#                         vusage_cleaned$nClones_PB/sum(vusage_cleaned$nClones_PB), 
#                         lower.tail = F))


vusage_PB <- total.apply %>% filter(source=="PB") %>% group_by(V, donor, kmer) %>% summarise(nCLones=n(), nUMIs=sum(count))
vusage_SF <- total.apply %>% filter(source=="SF") %>% group_by(V, donor, kmer) %>% summarise(nCLones=n(), nUMIs=sum(count))

vusage <- merge(vusage_PB, vusage_SF, by=c("V", "donor", "kmer"), all=T) 
#View(vusage)
vusage
vusage_cleaned <- vusage %>% group_by(donor) %>% mutate(nClones_PB=ifelse(is.na(nCLones.x)==T,min(nCLones.x,na.rm = T)/10,nCLones.x),nClones_SF=ifelse(is.na(nCLones.y)==T,min(nCLones.y,na.rm = T)/10,nCLones.y),nUMIs_SF=ifelse(is.na(nUMIs.x)==T,min(nUMIs.x,na.rm = T)/10,nUMIs.x),nUMIs_PB=ifelse(is.na(nUMIs.y)==T,min(nUMIs.y,na.rm = T)/10,nUMIs.y)) %>% ungroup() 
vusage_cleaned %<>% select(-nCLones.x, -nCLones.y, -nUMIs.x, -nUMIs.y)
#View(vusage_cleaned)
vusage_cleaned
vusage_cleaned %<>% mutate(nClones_PB=as.integer(nClones_PB), nClones_SF=as.integer(nClones_SF)) 
vusage_cleaned <- vusage_cleaned %>% 
  mutate(pbinom= pbinom(vusage_cleaned$nClones_SF, sum(vusage_cleaned$nClones_SF), 
                        vusage_cleaned$nClones_PB/sum(vusage_cleaned$nClones_PB), 
                        lower.tail = F)) %>% mutate(pbinom_adj = p.adjust(pbinom))

#View(vusage_cleaned)
vusage_cleaned

calculate_common_kmers <- function(n1, n2){
  t1 <- vusage_cleaned %>% filter(donor == n1 & pbinom_adj < 0.01)
  t2 <- vusage_cleaned %>% filter(donor == n2 & pbinom_adj < 0.01)
  result <- merge(t1, t2, by=c("kmer", "V"))
  #View(result)
  return(nrow(result))
}

#calculate_common_kmers("Dv", "Ali")

a <- combn(unique(paired_pb_sf_cd8$donor), 2, FUN = function (x) x <- c(x, toString(calculate_common_kmers(x[[1]], x[[2]]))) , simplify =F)
a
```

```{r}

m = matrix(c(0)*13*13, nrow=13, ncol=13)
m
colnames(m) <- unique(paired_pb_sf_cd8$donor)
rownames(m) <- unique(paired_pb_sf_cd8$donor)
m

# m["Dv", "Dv"]

for (i in a){
  m[i[1], i[2]] = i[3]
}
m

m <- as.data.frame(m)
m
head(m, n = 3)
```

```{r}
d <- dist(m, method = "euclidean") # distance matrix
fit <- hclust(d, method="ward") 
plot(fit) # display dendogram

groups <- cutree(fit, k=5) # cut tree into 5 clusters
# draw dendogram with red borders around the 5 clusters 
rect.hclust(fit, k=5, border="red")


```

```{r}

library(pvclust)
fit <- pvclust(m, method.hclust="ward",
               method.dist="euclidean")
plot(fit) # dendogram with p values
# add rectangles around groups highly supported by the data
pvrect(fit, alpha=.95)

```

```{r}
# K-Means Cluster Analysis
fit <- kmeans(m, 2) # 5 cluster solution
# get cluster means 
aggregate(m,by=list(fit$cluster),FUN=mean)
# append cluster assignment
m <- data.frame(m, fit$cluster)
m

```

```{r}
m$fit.cluster = NULL
m
```


```{r}
# # Model Based Clustering
library(mclust)
# fit <- Mclust(m)
# plot(fit) # plot results 
# summary(fit) # display the best model

```

```{r}
library(factoextra)
library(cluster)
library(factoextra)
dd <- daisy(m)
dd
round(as.matrix(dd)[1:13, 1:13], 2)

fviz_dist(dd)
```




```{r}

mydata <- scale(dd) 
fviz_nbclust(mydata, kmeans, method = "gap_stat")


set.seed(123) # for reproducibility
km.res <- kmeans(mydata, 2, nstart = 25)
# Visualize
fviz_cluster(km.res, data = mydata, palette = "jco",
             ggtheme = theme_minimal())


res.hc <- hclust(dist(mydata),  method = "ward.D2")
fviz_dend(res.hc, cex = 0.5, k = 4, palette = "jco") 

library(pheatmap)
pheatmap(t(mydata), cutree_cols = 4, kmeans_k = 2)

```



```{r}
# K-Means Clustering with 5 clusters
fit <- kmeans(mydata, 2)

# Cluster Plot against 1st 2 principal components

# vary parameters for most readable graph
library(cluster) 
clusplot(mydata, fit$cluster, color=TRUE, shade=TRUE, 
  	labels=2, lines=0)

# Centroid Plot against 1st 2 discriminant functions
library(fpc)
plotcluster(mydata, fit$cluster)



distance <- get_dist(dd)
fviz_dist(distance, gradient = list(low = "#00AFBB", mid = "white", high = "#FC4E07"))
```





```{r}
library(Biostrings)
require(ggplot2)
require(ggseqlogo)
library(msa)
library(seqRFLP)
library(RWebLogo)
require(ggplot2)
require(ggseqlogo)
library(DiffLogo)


# only  4 B27+

motive_table <- final_sure %>% 
  mutate(lenKmer=nchar(kmer)) %>% 
  filter(lenKmer==4) %>% 
  filter(V%in% HUMAN_TRBV) %>% 
  filter(N_B27 == 6) %>% 
  arrange(desc(fold_change_m))

motive_arr <- motive_table$kmer
motive_arr

selected_petterns <- Patients_common_table[grep(motive_arr[1], Patients_common_table$CDR3aa), ] %>% select(CDR3aa) %>% mutate(len=nchar(CDR3aa))
qplot(selected_petterns$len,
      geom="histogram")

df.fasta = dataframe2fas(selected_petterns, file="df.fasta")



for (i in motive_arr){Patients_common_table[grep(i, Patients_common_table$CDR3aa), ] %>% select(CDR3aa) %>%
    dataframe2fas(file= "./aa_seq/" %+% i %+% ".fasta")
}


```



```{r}

l = list()
j = 1
for (i in motive_arr[1]){ Patients_common_table[grep(i, Patients_common_table$CDR3aa), ] %>% 
    filter(V=="TRBV9") %>% 
    mutate(len=nchar(CDR3aa)) -> table
    all_len <- unique(table$len)
    
    plot_list <- lapply(all_len, function(x){print(x)
      table %>% filter(len==x) %>% 
        pull(CDR3aa) %>% AAStringSet() %>% 
        msa(method="ClustalOmega") %>% 
        getPwmFromAlignment() -> a
        plot <- ggplot() + geom_logo(a) + theme_logo()+
            ggtitle('Motive logo for CD8 V9 for ' %+% toString(length(a)) %+% ' sequences')+
           theme(text = element_text(size = 18),
           axis.title.x = element_blank(),
           axis.title.y = element_blank(),
           panel.grid.minor.x = element_blank(),
           panel.grid.major.x = element_blank())
      
    })
    l[[i]] = plot_list
    j = j + 1
    
}


l[[1]]
l


sequences <- selected_petterns$CDR3aa
sequences
string_set <- AAStringSet(sequences)
alignment <- msa(string_set, method="ClustalOmega")
tmp = sapply(alignment@unmasked, as.character)




ggplot() + geom_logo(tmp) + theme_logo()

```



```{r}
l = list()
j = 1
for (i in motive_arr[3]){ Patients_common_table[grep(i, Patients_common_table$CDR3aa), ] %>% 
    filter(V=="TRBV9") %>% 
    mutate(len=nchar(CDR3aa)) -> table
    all_len <- unique(table$len)
    
    plot_list <- lapply(all_len, function(x){print(x)
      table %>% filter(len==x) %>% 
        pull(CDR3aa) %>% AAStringSet() %>% 
        msa(method="ClustalOmega") %>% 
        getPwmFromAlignment() -> a
        plot <- ggplot() + geom_logo(a) + theme_logo()+
            ggtitle('Motive logo for CD8 V9 for ' %+% toString(length(a)) %+% ' sequences')+
           theme(text = element_text(size = 18),
           axis.title.x = element_blank(),
           axis.title.y = element_blank(),
           panel.grid.minor.x = element_blank(),
           panel.grid.major.x = element_blank())
      
    })
    l[[i]] = plot_list
    j = j + 1
    
}


l[[1]]
l

ggseqlogo(tmp, ncol=4)
```



```{r}
l = list()
j = 1
for (i in motive_arr[2]){ Patients_common_table[grep(i, Patients_common_table$CDR3aa), ] %>% 
    filter(V=="TRBV9") %>% 
    mutate(len=nchar(CDR3aa)) -> table
    all_len <- unique(table$len)
    
    plot_list <- lapply(all_len, function(x){print(x)
      table %>% filter(len==x) %>% 
        pull(CDR3aa)
    })
    l[[i]] = plot_list
    j = j + 1
    
}

ggseqlogo(l[[1]], ncol=4)
```



```{r}
# only  4 B27+

motive_table <- final_sure %>% 
  mutate(lenKmer=nchar(kmer)) %>% 
  filter(lenKmer==4) %>% 
  filter(V%in% HUMAN_TRBV) %>% 
  filter(N_B38 == 3) %>% 
  arrange(desc(fold_change_m))

motive_arr <- motive_table$kmer
motive_arr



l = list()
j = 1
for (i in motive_arr[1]){ Patients_common_table[grep(i, Patients_common_table$CDR3aa), ] %>% 
    filter(V=="TRBV9") %>% 
    mutate(len=nchar(CDR3aa)) -> table
    all_len <- unique(table$len)
    
    plot_list <- lapply(all_len, function(x){print(x)
      table %>% filter(len==x) %>% 
        pull(CDR3aa)
    })
    l[[i]] = plot_list
    j = j + 1
    
}

ggseqlogo(l[[1]], ncol=4)
```



